File: CS/Server/BallGameServer/ClientHandler.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Sockets;
using System.Threading;

namespace BallGameServer
{
    public class ClientHandler
    {
        public static readonly Object sync = new object();
        public static HashSet<int> inGamePlayerIds = new HashSet<int>();
        private static HashSet<int> allPlayerIds = new HashSet<int>();
        public BlockingCollection<String> responseQueue = new BlockingCollection<String>();
        public StreamReader reader;
        public int playerId;
        private StreamWriter writer;
        private Thread responseControllerThread;

        public ClientHandler(TcpClient tcpClient)
        {
            Stream stream = tcpClient.GetStream();
            reader = new StreamReader(stream);
            writer = new StreamWriter(stream);
            playerId = GeneratePlayerId();
        }

        private int GeneratePlayerId()
        {
            int ID_BOUND = 10000;
            int playerId = (new Random()).Next(ID_BOUND);
            while (allPlayerIds.Contains(playerId))
            {
                playerId = (new Random()).Next(ID_BOUND);
            }
            allPlayerIds.Add(playerId);
            inGamePlayerIds.Add(playerId);
            return playerId;
        }

        private void BroadcastMessage(String message)
        {

            foreach (ClientHandler client in ServerProgram.clients)
            {
                try
                {
                    client.writer.WriteLine(message);
                    client.writer.Flush();
                }
                catch (IOException)
                {
                    // Client has disconnected
                }
            }

        }

        private void PrintAllPlayers()
        {
            Console.WriteLine("In-game players:");
            foreach (int id in inGamePlayerIds)
            {
                Console.WriteLine(id);
            }
        }

        private int GetNextAvailablePlayer()
        {
            int nextPlayer = 0;
            foreach (int id in inGamePlayerIds)
            {
                if (id != playerId)
                {
                    nextPlayer = id;
                    break;
                }
            }
            return nextPlayer;
        }

        private void RemovePlayerFromInGameSet(int playerId)
        {
            inGamePlayerIds.Remove(playerId);
        }

        public void PlayerDisconnected()
        {
            BroadcastMessage(playerId + " has disconnected#addBroadcast:broadcast");
            BroadcastMessage(playerId + "#removeId:broadcast");
            Console.WriteLine("\nPlayer " + playerId + " has disconnected");

            if (Ball.GetBallHolder() == playerId)  // Disconnecting player has the ball
            {
                if (inGamePlayerIds.Count() > 1)  // There is someone to pass ball to
                {
                    int newBallHolder = GetNextAvailablePlayer();
                    Ball.SetBallHolder(newBallHolder);
                    RemovePlayerFromInGameSet(playerId);
                    PrintAllPlayers();
                    BroadcastMessage(playerId + " disconnected. New ball holder is " + newBallHolder + "#addBroadcast:broadcast");
                    BroadcastMessage(newBallHolder + "#setBallHolder:broadcast");
                    Console.WriteLine("\nBall holder (" + playerId + ") has left the game, ball has been automatically passed to " + newBallHolder);
                }
                else  // There isn't someone to pass ball to
                {
                    Ball.SetBallHolder(-1);
                    RemovePlayerFromInGameSet(playerId);
                    PrintAllPlayers();
                    Console.WriteLine("\nLast player (" + playerId + ") has left, ball will be automatically passed to the next player that joins");
                }
            }
            else  // Disconnecting player does not have ball
            {
                RemovePlayerFromInGameSet(playerId);
                PrintAllPlayers();
            }

            ServerProgram.clients.Remove(this);
            reader.Close();
            writer.Close();
        }

        public void Run()
        {
            writer.WriteLine(playerId + ":individual");
            writer.Flush();

            Console.WriteLine("\nPlayer " + playerId + " has joined the game");
            PrintAllPlayers();

            writer.WriteLine("SUCCESS:individual");
            writer.Flush();
            BroadcastMessage("Player " + playerId + " has joined the game#addBroadcast:broadcast");
            BroadcastMessage(playerId + "#addId:broadcast");

            if (Ball.GetBallHolder() == -1)
            {
                Console.WriteLine("\n" + playerId + " has been passed the ball by the server");
                Ball.SetBallHolder(playerId);
            }

            responseControllerThread = new Thread(new ResponseController(this).Run);
            responseControllerThread.Start();

            while (true)
            {
                String clientRequest = responseQueue.Take();
                switch (clientRequest)
                {
                    case "ball_holder":
                        writer.WriteLine(Ball.GetBallHolder() + ":individual");
                        writer.Flush();
                        break;
                    case "pass_ball":
                        int newBallHolder = int.Parse(responseQueue.Take());
                        Console.WriteLine("\nBall passed: " + Ball.GetBallHolder() + " -> " + newBallHolder);
                        BroadcastMessage("Ball passed " + Ball.GetBallHolder() + " -> " + newBallHolder + "#addBroadcast:broadcast");
                        BroadcastMessage(newBallHolder + "#setBallHolder:broadcast");
                        Ball.SetBallHolder(newBallHolder);
                        break;
                    case "get_players":
                        writer.WriteLine(inGamePlayerIds.Count() + ":individual");
                        writer.Flush();
                        foreach (int id in inGamePlayerIds)
                        {
                            writer.WriteLine(id + ":individual");
                            writer.Flush();
                        }
                        break;
                }
            }
        }

        public override string ToString()
        {
            return "Client:" + playerId;
        }
    }
}

File: Java/Server/src/ClientHandler.java
import java.io.IOException;
import java.io.PrintWriter;
import java.net.Socket;
import java.util.HashSet;
import java.util.Random;
import java.util.Scanner;
import java.util.Set;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ClientHandler implements Runnable {

    final static Object sync = new Object();
    static Set<Integer> inGamePlayerIds = new HashSet<>();
    private static Set<Integer> allPlayerIds = new HashSet<>();
    BlockingQueue<String> responseQueue = new LinkedBlockingQueue<>();
    Scanner reader;
    int playerId;
    private PrintWriter writer;

    ClientHandler(Socket socket) throws IOException {
        reader = new Scanner(socket.getInputStream());
        writer = new PrintWriter(socket.getOutputStream(), true);
        playerId = generatePlayerId();
    }

    private int generatePlayerId() {
        final int ID_BOUND = 10000;
        int playerId = (new Random()).nextInt(ID_BOUND);
        while (allPlayerIds.contains(playerId)) {
            playerId = (new Random()).nextInt(ID_BOUND);
        }
        allPlayerIds.add(playerId);
        inGamePlayerIds.add(playerId);
        return playerId;
    }

    private void broadcastMessage(String message) {
        for (ClientHandler client : ServerProgram.clients) {
            client.writer.println(message);
        }
    }

    private void printAllPlayers() {
        System.out.println("In-game players:");
        for (int id : inGamePlayerIds) {
            System.out.println(id);
        }
    }

    private int getNextAvailablePlayer() {
        int nextPlayer = 0;
        for (int id : inGamePlayerIds) {
            if (id != playerId) {
                nextPlayer = id;
                break;
            }
        }
        return nextPlayer;
    }

    private void removePlayerFromInGameSet(int playerId) {
        inGamePlayerIds.removeIf(id -> id.equals(playerId));
    }

    void playerDisconnected() {
        broadcastMessage(playerId + " has disconnected#addBroadcast:broadcast");
        broadcastMessage(playerId + "#removeId:broadcast");
        System.out.println("\nPlayer " + playerId + " has disconnected");

        if (Ball.getBallHolder() == playerId) {  // Disconnecting player has the ball
            if (inGamePlayerIds.size() > 1) {  // There is someone to pass ball to
                int newBallHolder = getNextAvailablePlayer();
                Ball.setBallHolder(newBallHolder);
                removePlayerFromInGameSet(playerId);
                printAllPlayers();
                broadcastMessage(playerId + " disconnected. New ball holder is " + newBallHolder + "#addBroadcast:broadcast");
                broadcastMessage(newBallHolder + "#setBallHolder:broadcast");
                System.out.println("\nBall holder (" + playerId + ") has left the game, ball has been automatically passed to " + newBallHolder);
            } else {  // There isn't someone to pass ball to
                Ball.setBallHolder(-1);
                removePlayerFromInGameSet(playerId);
                printAllPlayers();
                System.out.println("\nLast player (" + playerId + ") has left, ball will be automatically passed to the next player that joins");
            }
        } else {  // Disconnecting player does not have ball
            removePlayerFromInGameSet(playerId);
            printAllPlayers();
        }

        ServerProgram.clients.remove(this);
        reader.close();
        writer.close();
    }

    @Override
    @SuppressWarnings("InfiniteLoopStatement")
    public void run() {
        try {
            writer.println(playerId + ":individual");

            System.out.println("\nPlayer " + playerId + " has joined the game");
            printAllPlayers();

            writer.println("SUCCESS:individual");
            broadcastMessage("Player " + playerId + " has joined the game#addBroadcast:broadcast");
            broadcastMessage(playerId + "#addId:broadcast");

            if (Ball.getBallHolder() == -1) {
                System.out.println("\n" + playerId + " has been passed the ball by the server");
                Ball.setBallHolder(playerId);
            }

            new Thread(new ResponseController(this)).start();

            while (true) {
                String clientRequest = responseQueue.take();
                switch (clientRequest) {
                    case "ball_holder":
                        writer.println(Ball.getBallHolder() + ":individual");
                        break;
                    case "pass_ball":
                        int newBallHolder = Integer.parseInt(responseQueue.take());
                        System.out.println("\nBall passed: " + Ball.getBallHolder() + " -> " + newBallHolder);
                        broadcastMessage("Ball passed " + Ball.getBallHolder() + " -> " + newBallHolder + "#addBroadcast:broadcast");
                        broadcastMessage(newBallHolder + "#setBallHolder:broadcast");
                        Ball.setBallHolder(newBallHolder);
                        break;
                    case "get_players":
                        writer.println(inGamePlayerIds.size() + ":individual");
                        for (int id : inGamePlayerIds) {
                            writer.println(id + ":individual");
                        }
                        break;
                }
            }
        } catch (InterruptedException e) {
            System.err.println("Error while removing item from response queue\n" + e.getMessage());
        }
    }
}
